### ISSUE TITLE
Grafana v9 is trying to merge times values axes when using mulitple metrics and create 0 value if need be

### ISSUE THREAD

**Issue by lbrachet** [grafana/grafana#65319]
Hello,

I am migrating from Grafana v7.3.7 to Grafana v9.4.3.

**What happened**:

I encounter a blocking problem during my migration concerning the use of a Graph panel containing only 1 query A but returning a dataset with two metrics, server_1 & server_2.

I insist on the fact that it is only one query A returning the 2 metrics, and not two metrics A and B returning respectively the metrics of server_1 and server_2 because in this case I have no problem.

![grafana-v9](https://user-images.githubusercontent.com/121950145/227580156-e995521c-8e47-4497-92ea-1589a791a179.png)

The reason is that grafana v9 is changing ma dataset by that :
![grafana-dataset-v9](https://user-images.githubusercontent.com/121950145/227580404-e9a5fda3-a45c-4129-91bc-8f87c8392bf7.png)

**What you expected to happen**:

I want this:
![grafana-v7](https://user-images.githubusercontent.com/121950145/227580193-f8c3fb4c-8bf2-43f0-a532-d823f6875ffc.png)

**How to reproduce it (as minimally and precisely as possible)**:

My source is mssql which returns this dataset:

![grafana-dataset-v7](https://user-images.githubusercontent.com/121950145/227580099-d9490006-34e4-4d83-9b03-9b1af13d6cb6.png)

**Anything else we need to know?**:

So it is trying to find a value for each metric and for each time that he will find the dataset, and is virtualing creating a measure with a 0 value for each correponding missing matchs.

This behavior is unwanted and can not be changed with connect as null…

How can I tell Grafana to not add this 0 values?

**Environment**:
- Grafana plugins: briangann-datatable-panel / btplc-alarm-box-panel / grafana-polystat-panel / grafana-simple-json-datasource / xginn8-pagerduty-datasource / vonage-status-panel / jdbranham-diagram-panel / briangann-gauge-panel / novatec-sdg-panel / yesoreyeram-boomtable-panel

**Comment by leeoniya** [comment]
please attach a debug snapshot (it will have the raw response data embedded in it). these snapshots can randomize sensitive data which you can review prior to posting.

https://grafana.com/docs/grafana/latest/whatsnew/whats-new-in-v9-2/#panel-help-menu

![image](https://user-images.githubusercontent.com/43234/200391751-ca1c338a-460b-46de-b881-87ea81e55f5d.png)

**Comment by leeoniya** [comment]
i am fairly confident that what you're looking to do with the data you're showing is `Partition by values`:

https://grafana.com/docs/grafana/latest/panels-visualizations/query-transform-data/transform-data/#partition-by-values

it would still be good to see how the same queries & data behaves in Graph (old) vs TimeSeries so we can properly handle this in our auto-migration code.

**Comment by jbobin-betclic** [comment]
Hello, 

Here are the export from the v9. 

`
| Key | Value |
|--|--|
| Panel | graph @ 9.4.3 |
| Grafana | 9.4.3 (cf0a135595) // Open Source |
<details><summary>Panel debug snapshot dashboard</summary>

```json
{
  "panels": [
    {
      "datasource": {
        "type": "grafana",
        "uid": "grafana"
      },
      "aliasColors": {},
      "bars": false,
      "dashLength": 10,
      "dashes": false,
      "description": "source: view_fact_performance_counter",
      "fill": 0,
      "fillGradient": 0,
      "gridPos": {
        "h": 13,
        "w": 15,
        "x": 0,
        "y": 0
      },
      "hiddenSeries": false,
      "id": 2,
      "legend": {
        "alignAsTable": true,
        "avg": true,
        "current": false,
        "max": true,
        "min": true,
        "show": false,
        "sort": "total",
        "sortDesc": true,
        "total": true,
        "values": true
      },
      "lines": true,
      "linewidth": 2,
      "links": [],
      "nullPointMode": "null",
      "options": {
        "alertThreshold": true
      },
      "percentage": false,
      "pluginVersion": "9.4.3",
      "pointradius": 5,
      "points": false,
      "renderer": "flot",
      "seriesOverrides": [],
      "spaceLength": 10,
      "stack": false,
      "steppedLine": false,
      "targets": [
        {
          "refId": "A",
          "datasource": {
            "type": "grafana",
            "uid": "grafana"
          },
          "queryType": "snapshot",
          "snapshot": [
            {
              "schema": {
                "refId": "A",
                "meta": {
                  "type": "timeseries-wide",
                  "executedQueryString": "xxx"
                },
                "fields": [
                  {
                    "name": "Time",
                    "type": "time",
                    "typeInfo": {
                      "frame": "time.Time"
                    },
                    "config": {}
                  },
                  {
                    "name": "server_2",
                    "type": "number",
                    "typeInfo": {
                      "frame": "float64"
                    },
                    "config": {}
                  },
                  {
                    "name": "server_1",
                    "type": "number",
                    "typeInfo": {
                      "frame": "float64"
                    },
                    "config": {}
                  }
                ]
              },
              "data": {
                "values": [
                  [
                    1679587214000,
                    1679587222000,
                    1679587234000,
                    1679587236000,
                    1679587252000,
                    1679587256000,
                    1679587274000,
                    1679587280000,
                    1679587294000,
                    1679587296000,
                    1679587306000,
                    1679587316000,
                    1679587326000,
                    1679587336000
                  ],
                  [
                    0,
                    16,
                    0,
                    7,
                    0,
                    8,
                    0,
                    12,
                    0,
                    6,
                    0,
                    8,
                    0,
                    14
                  ],
                  [
                    38,
                    0,
                    32,
                    0,
                    27,
                    0,
                    29,
                    0,
                    35,
                    0,
                    21,
                    0,
                    33,
                    0
                  ]
                ]
              }
            }
          ]
        }
      ],
      "thresholds": [],
      "timeRegions": [],
      "title": "Reproduced with embedded data",
      "tooltip": {
        "shared": true,
        "sort": 0,
        "value_type": "individual"
      },
      "type": "graph",
      "xaxis": {
        "mode": "time",
        "show": true,
        "values": [],
        "name": null,
        "buckets": null
      },
      "yaxes": [
        {
          "$$hashKey": "object:1592",
          "format": "short",
          "logBase": 1,
          "show": true
        },
        {
          "$$hashKey": "object:1593",
          "format": "short",
          "logBase": 1,
          "show": true
        }
      ],
      "yaxis": {
        "align": false
      },
      "timeFrom": null,
      "timeShift": null
    },
    {
      "gridPos": {
        "h": 7,
        "w": 9,
        "x": 15,
        "y": 0
      },
      "id": 5,
      "options": {
        "content": "<table width=\"100%\">\n    <tr>\n      <th width=\"2%\">Panel</th>\n      <td >graph @ 9.4.3</td>\n    </tr>\n    <tr>\n      <th>Queries</th>\n      <td>A[mssql]</td>\n    </tr>\n    \n    <tr><th>Data</th><td> 1 frames, 3 fields, 14 rows </td></tr>\n    \n    <tr>\n      <th>Grafana</th>\n      <td>9.4.3 (cf0a135595) // Open Source</td>\n    </tr>\n  </table>",
        "mode": "html"
      },
      "title": "Debug info",
      "type": "text"
    },
    {
      "id": 6,
      "title": "Original Panel JSON",
      "type": "text",
      "gridPos": {
        "h": 13,
        "w": 9,
        "x": 15,
        "y": 7
      },
      "options": {
        "content": "{\n  \"datasource\": {\n    \"type\": \"mssql\",\n    \"uid\": \"000000055\"\n  },\n  \"aliasColors\": {},\n  \"bars\": false,\n  \"dashLength\": 10,\n  \"dashes\": false,\n  \"description\": \"source: view_fact_performance_counter\",\n  \"fill\": 0,\n  \"fillGradient\": 0,\n  \"gridPos\": {\n    \"h\": 4,\n    \"w\": 4,\n    \"x\": 10,\n    \"y\": 1\n  },\n  \"hiddenSeries\": false,\n  \"id\": 364,\n  \"legend\": {\n    \"alignAsTable\": true,\n    \"avg\": true,\n    \"current\": false,\n    \"max\": true,\n    \"min\": true,\n    \"show\": false,\n    \"sort\": \"total\",\n    \"sortDesc\": true,\n    \"total\": true,\n    \"values\": true\n  },\n  \"lines\": true,\n  \"linewidth\": 2,\n  \"links\": [],\n  \"nullPointMode\": \"null\",\n  \"options\": {\n    \"alertThreshold\": true\n  },\n  \"percentage\": false,\n  \"pluginVersion\": \"9.4.3\",\n  \"pointradius\": 5,\n  \"points\": false,\n  \"renderer\": \"flot\",\n  \"seriesOverrides\": [],\n  \"spaceLength\": 10,\n  \"stack\": false,\n  \"steppedLine\": false,\n  \"targets\": [\n    {\n      \"alias\": \"\",\n      \"datasource\": {\n        \"type\": \"mssql\",\n        \"uid\": \"000000055\"\n      },\n      \"format\": \"time_series\",\n      \"rawSql\": \"DECLARE\\r\\n  @from datetime = $__timeFrom(),\\r\\n  @to datetime = $__timeTo()\\r\\n\\r\\nDECLARE @Server NVARCHAR(MAX) = '${Server:csv}' \\r\\n xxx \\r\\n\",\n      \"refId\": \"A\"\n    }\n  ],\n  \"thresholds\": [],\n  \"timeRegions\": [],\n  \"title\": \"Batch Requests/sec\",\n  \"tooltip\": {\n    \"shared\": true,\n    \"sort\": 0,\n    \"value_type\": \"individual\"\n  },\n  \"type\": \"graph\",\n  \"xaxis\": {\n    \"mode\": \"time\",\n    \"show\": true,\n    \"values\": [],\n    \"name\": null,\n    \"buckets\": null\n  },\n  \"yaxes\": [\n    {\n      \"$$hashKey\": \"object:1592\",\n      \"format\": \"short\",\n      \"logBase\": 1,\n      \"show\": true\n    },\n    {\n      \"$$hashKey\": \"object:1593\",\n      \"format\": \"short\",\n      \"logBase\": 1,\n      \"show\": true\n    }\n  ],\n  \"yaxis\": {\n    \"align\": false\n  },\n  \"timeFrom\": null,\n  \"timeShift\": null\n}",
        "mode": "code",
        "code": {
          "language": "json",
          "showLineNumbers": true,
          "showMiniMap": true
        }
      }
    },
    {
      "id": 3,
      "title": "Data from panel above",
      "type": "table",
      "datasource": {
        "type": "datasource",
        "uid": "-- Dashboard --"
      },
      "gridPos": {
        "h": 7,
        "w": 15,
        "x": 0,
        "y": 13
      },
      "options": {
        "showTypeIcons": true
      },
      "targets": [
        {
          "datasource": {
            "type": "datasource",
            "uid": "-- Dashboard --"
          },
          "panelId": 2,
          "withTransforms": true,
          "refId": "A"
        }
      ]
    }
  ],
  "schemaVersion": 37,
  "title": "Debug: Batch Requests/sec // 2023-03-25 17:21:04",
  "tags": [
    "debug",
    "debug-graph"
  ],
  "time": {
    "from": "2023-03-23T16:00:00.000Z",
    "to": "2023-03-23T16:02:18.271Z"
  }
}
```
</details>
`

**Comment by jbobin-betclic** [comment]
> i'm fairly confident that what you're looking to do with the data you're showing is `Partition by values`:
> 
> https://grafana.com/docs/grafana/latest/panels-visualizations/query-transform-data/transform-data/#partition-by-values
> 
> it would still be good to see how the same queries & data behaves in Graph (old) vs TimeSeries so we can properly handle this in our auto-migration code.

Transformation `Partition by values` is not working in our case. 

Our problem is that v7 produce a data with null values when not finding a value for a same time, will v9 produce a 0 value, which is not a good thing from my point of view. 

Inspecting data for the v7 version pro with formatted_data:no produces 
![image](https://user-images.githubusercontent.com/38936113/227730173-e50b27c5-bcaf-4ea9-9cb0-ae2c3739a6a6.png)

while inspecting data for the v9 version with formatted_data: no produces
![image](https://user-images.githubusercontent.com/38936113/227730289-a92e86f3-525c-4c34-971d-bebe1d104ca6.png)

**Comment by leeoniya** [comment]
thanks for the info, passing this along to @grafana/grafana-bi-squad since it appears to be a change in the mssql datasource response.

with such a big version jump it would be useful to pinpoint which release changed the response format to return zeros.
it may have been https://github.com/grafana/grafana/pull/32257, which landed in **8.0.0-beta1**.  you can maybe try some intermediate versions starting with latest 8.0.x, 8.1.x, and so on: https://grafana.com/grafana/download?edition=oss

**EDIT:** i think this was already reported, and the solution is here: https://github.com/grafana/grafana/issues/40680#issuecomment-949374769
